File: companion_core/features.py
from companion_core.filtering import Ema, Trend
from companion_core.types import Baseline, HealthFeatures, SensorReading, SensorState


class FeatureExtractor:
    def __init__(self):
        self.hr_ema = Ema()
        self.spo2_ema = Ema()
        self.temp_ema = Ema()
        self.amb_ema = Ema()
        self.hum_ema = Ema()
        self.mq_ema = Ema()
        self.hr_tr = Trend()
        self.spo2_tr = Trend()
        self.temp_tr = Trend()
        self._persist_hits = 0

    def extract(self, r: SensorReading, baseline: Baseline, prev_abnormal: bool) -> HealthFeatures:
        f = HealthFeatures()
        f.signal_quality = r.ppg_quality

        if r.max30102_state == SensorState.OK:
            f.hr = self.hr_ema.update(r.hr_bpm)
            f.spo2 = self.spo2_ema.update(r.spo2_pct)
            f.hr_trend = self.hr_tr.update(f.hr)
            f.spo2_trend = self.spo2_tr.update(f.spo2)
        if r.mlx_state == SensorState.OK:
            f.temperature = self.temp_ema.update(r.object_temp_c)
            f.temperature_trend = self.temp_tr.update(f.temperature)
        if r.dht_state == SensorState.OK:
            f.ambient_temp = self.amb_ema.update(r.dht_temp_c)
            f.humidity = self.hum_ema.update(r.humidity_pct)
        if r.mq_state in (SensorState.OK,):
            f.mq135_relative = self.mq_ema.update(r.mq135_relative)

        if f.hr is not None:
            f.hr_dev = (f.hr - baseline.resting_hr) / max(baseline.hr_range, 4.0)
        if f.spo2 is not None:
            f.spo2_dev = (baseline.typical_spo2 - f.spo2) / 4.0
        if f.temperature is not None:
            f.temp_dev = (f.temperature - baseline.typical_temp) / 1.2

        if prev_abnormal:
            self._persist_hits += 1
        else:
            self._persist_hits = max(0, self._persist_hits - 1)
        f.persistence = min(1.0, self._persist_hits / 6.0)

        f.valid = (f.hr is not None and f.spo2 is not None and f.temperature is not None
                   and f.signal_quality >= 0.45)
        return f

File: firmware/src/processing/features.cpp
#include "companion/features.h"
#include "companion/config.h"
#include <math.h>
#include <string.h>

void companion_features_init(companion_feature_state_t *s) {
  memset(s, 0, sizeof(*s));
  companion_ema_init(&s->hr, COMPANION_EMA_ALPHA);
  companion_ema_init(&s->spo2, COMPANION_EMA_ALPHA);
  companion_ema_init(&s->temp, COMPANION_EMA_ALPHA);
  companion_ema_init(&s->amb, COMPANION_EMA_ALPHA);
  companion_ema_init(&s->hum, COMPANION_EMA_ALPHA);
  companion_ema_init(&s->mq, 0.15f);
}

void companion_features_extract(
    companion_feature_state_t *s,
    const companion_reading_t *r,
    const companion_baseline_t *b,
    uint8_t prev_abnormal,
    companion_features_t *out) {
  memset(out, 0, sizeof(*out));
  out->signal_quality = r->ppg_quality;

  if (r->max_state == COMP_OK && r->hr_valid) {
    out->hr = companion_ema_update(&s->hr, r->hr, 1);
    if (s->hr_prev_ok) out->hr_trend = out->hr - s->hr_prev;
    s->hr_prev = out->hr;
    s->hr_prev_ok = 1;
  }
  if (r->max_state == COMP_OK && r->spo2_valid) {
    out->spo2 = companion_ema_update(&s->spo2, r->spo2, 1);
    if (s->spo2_prev_ok) out->spo2_trend = out->spo2 - s->spo2_prev;
    s->spo2_prev = out->spo2;
    s->spo2_prev_ok = 1;
  }
  if (r->mlx_state == COMP_OK && r->temp_valid) {
    out->temperature = companion_ema_update(&s->temp, r->object_temp_c, 1);
    if (s->temp_prev_ok) out->temperature_trend = out->temperature - s->temp_prev;
    s->temp_prev = out->temperature;
    s->temp_prev_ok = 1;
  }
  if (r->dht_valid) {
    out->ambient_temp = companion_ema_update(&s->amb, r->dht_temp_c, 1);
    out->humidity = companion_ema_update(&s->hum, r->humidity, 1);
  }
  if (r->mq_state == COMP_OK) {
    out->mq135_relative = companion_ema_update(&s->mq, r->mq135_relative, 1);
  }

  if (out->hr > 0) out->hr_dev = (out->hr - b->resting_hr) / (b->hr_range > 4.f ? b->hr_range : 4.f);
  if (out->spo2 > 0) out->spo2_dev = (b->typical_spo2 - out->spo2) / 4.f;
  if (out->temperature > 0) out->temp_dev = (out->temperature - b->typical_temp) / 1.2f;

  if (prev_abnormal) {
    if (s->persist_hits < 100) s->persist_hits++;
  } else if (s->persist_hits > 0) {
    s->persist_hits--;
  }
  out->persistence = s->persist_hits / 6.f;
  if (out->persistence > 1.f) out->persistence = 1.f;

  out->valid = (out->hr > 0 && out->spo2 > 0 && r->ppg_quality >= COMPANION_SIGNAL_Q_MIN) ? 1 : 0;
#if !COMPANION_CAL_MLX_OPTIONAL
  if (out->temperature <= 0) out->valid = 0;
#endif
}

void companion_features_vector(const companion_features_t *f, float *x, int n) {
  if (n < 14) return;
  x[0] = f->hr;
  x[1] = f->spo2;
  x[2] = f->temperature;
  x[3] = f->hr_trend;
  x[4] = f->spo2_trend;
  x[5] = f->temperature_trend;
  x[6] = f->signal_quality;
  x[7] = f->hr_dev;
  x[8] = f->spo2_dev;
  x[9] = f->temp_dev;
  x[10] = f->persistence;
  x[11] = f->ambient_temp;
  x[12] = f->humidity;
  x[13] = f->mq135_relative;
}

void companion_flood_vector(float rain24, float *x, int n) {
  if (n < 9) return;
  memset(x, 0, sizeof(float) * (size_t)n);
  x[4] = rain24;
  x[6] = rain24;
}
